### ISSUE TITLE
Add the option to cancel sagas on location change and to choose buffer size and behavior 

### ISSUE THREAD

**Issue by victorchabbert** [jfairbank/redux-saga-router#6]
Like the title says, I think it would be useful to have an option to cancel sagas on a location change and/or to have control of the saga buffer globally and per route. This could probably be split into two issues. Thoughts ?

About the cancellation problem, I've explored the source already and saw that you do not keep a reference of the previous value which would have helped the cancellation of the value if it's a saga. How would you go about that ?

I'd like to help you figure it out and why not make my first PR !

Cheers :)

**Comment by jfairbank** [comment]
Hey, @victorchabbert!

Initially, I'm a little confused how this would be used. Which saga exactly are you wanting to cancel? The actual router saga, or the route sagas? The route sagas currently aren't forked, so we don't have tasks we can cancel. Are you able to provide a concrete example where this feature would be useful and how you envision being able to cancel saga(s) with some example code. Not saying what you're proposing couldn't be useful; I just don't fully understand it and how it would be used.

Thanks!

**Comment by victorchabbert** [comment]
Sorry for not being clear in the first place, I'll try to answer best all of your questions.
I would like to have more control over the route sagas.
Currently, my route sagas forks off a sub-saga located in an other file. That sub-saga lives forever until it's either canceled or it naturally finishes. As such, I manually cancel it on a LOCATION_CHANGE like so:
(Not cancelling the saga causes a buffer overflow)
```
const routes = {
  *'/'() {
    const saga = yield fork(longLivingSaga)
    yield take(LOCATION_CHANGE)
    yield cancel(saga)
  },
  ...
}
```
This code can become cumbersome with multiple routes.

It would be nice if we can have an api like the following:
```
const routes = [
  // for a referenced saga
  {
    path: '/',
    saga: longLivingSaga,
    method: 'fork',
    cancel: true
  },
  // for an inline saga
  {
    path: '/test',
    *'saga'() {
      ...
    }
  }, // method defaults to 'call'
  ...
]
```

This may not be the best solution but it's a start.

A concrete use case of this would be an app that heavily uses websockets in sagas. For a page, it subscribes to specific channels and unsubscribes by canceling the parent saga when the page is changed to avoid receiving and handling unnecessary information.

I hope it's clearer :)
Cheers!

**Comment by jfairbank** [comment]
Thanks for the clarification. Being able to fork inside a route saga and cancel the task sounds useful.

Honestly, it might make sense to just default to `fork` instead of `call`. I'm not sure why I went with `call` initially. I'm not sure if this would result in a breaking change for anyone, though. I would strongly prefer to avoid adding extra options to the routes if at all possible. Worst case is to just change it to default to `fork` and bump major version number to be safe.

Initially, I'm envisioning that we `fork` the route saga and save the task. Whenever a route changes, cancel the current task. That gives the route sagas an opportunity to use the `try..finally` pattern with [`cancelled`](https://redux-saga.github.io/redux-saga/docs/api/index.html#cancelled) to do any cleanup as well.

Does that sound like it would cover your use case better?

**Comment by victorchabbert** [comment]
It does however I feel like cancelling the current task by default closes off some current use cases.
For example, an app that uses route sagas for data fetching would want to have the execution of its saga to complete even if the location has changed.

To avoid this, I propose something like the following:
```
const options = {
  cancel: true
}

yield* router(history, routes, options);
```

If no options object is passed, it defaults to
```
const options = {
  cancel: false
}
```

From this, we could also add more options like choosing the effect creator (fork, spawn, call, ...), etc.

What do you think ?

**Comment by jfairbank** [comment]
Hmm, I'm not sure that that use case is compelling enough. This is no different than a multipage app with a normal server as a router. If there is an AJAX request in flight, and you click on a link, then that request should be cancelled as you go to the next page. Why would you need data to finish fetching when you want to go to some other route?

Honestly, I'm not in favor of supporting `call` any more either. In hindsight, I never should have used `call`, and I don't think it is a good idea as an option. Using `call` could block other route sagas from running immediately if the current route saga is taking too long. This router should really model the boss-worker pattern where the router saga is just listening and then handing off work to "threads", i.e. forked sagas.

I really think moving forward that `fork` needs to be the default and that no additional options are needed yet. That keeps the API the simplest. If we immediately add options without knowing if they're really needed, then that's just extra code and complexity to support. If options are needed in the future, then it can be revisited, but I'm not convinced enough yet.

Thank you for raising the original issue, though! I think this generated good dialogue and helped me realize that `call` wasn't a good idea originally. If you would like to implement `fork` as the default with task cancellation, I'd still love to give you an opportunity to submit a PR.

**Comment by victorchabbert** [comment]
Alright, I understand and I agree with you.

Thanks, I'll get work!

**Comment by jfairbank** [comment]
Implemented in #7.
